### ISSUE TITLE
postgres dumps hang when routed through linkerd

### ISSUE THREAD

**Issue by sjahl** [linkerd/linkerd2#5539]
### What is the issue?

I've got a postgres 12.1 database, and a python (django 3.1.3) web application running in two separate pods in my GKE cluster. Normal (mostly very quick) queries between the two pods are fine when meshed through linkerd. However, when trying to dump the database (with `pg_dump` when exec'ing into the web app pod), the dump hangs partway through when traffic is proxied by linkerd. I've observed this behavior with stable-2.9.1 and port 5432 **not** skipped. I also tried the edge-21.1.1 release, and made port 5432 `opaque`, and saw the behavior there too. In both releases, adding port 5432 to skipped ports allows database dumps to continue smoothly.

The amount of data that gets dumped seems inconsistent -- sometimes I get 100MBs from pg_dump before it hangs, other times it's only a MB or two.

### How can it be reproduced?

I can reproduce the behavior in my environment, with this sample database: https://github.com/anthonydb/practical-sql/tree/master/Chapter_11

If I load that CSV into postgres using the steps in the Chapter_11.sql file, then exec into another pod and try to dump the data, I see the hang. In my case, I created a database called linkerd_debug to put the sample table into, and dump like this: `/usr/bin/pg_dump -U postgres --host postgres linkerd_debug` (`postgres` is the name of the k8s service where my psql database pod lives)

### Logs, error output, etc

I've uploaded some tcpdump output (filtered on port 5432) of a dump that's hanging: https://gist.github.com/sjahl/52ae9e3fff78213b2697e96e382e2b6b

In the gist, you can see that packets abruptly stop at `09:29:37.235702`, at which point I waited about a minute and then ctrl-C'd the dump.

#### `linkerd check` output

```text
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ controller pod is running
√ can initialize the client
√ can query the control plane API

linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
√ control plane CustomResourceDefinitions exist
√ control plane MutatingWebhookConfigurations exist
√ control plane ValidatingWebhookConfigurations exist
√ control plane PodSecurityPolicies exist

linkerd-identity
----------------
√ certificate config is valid
√ trust anchors are using supported crypto algorithm
√ trust anchors are within their validity period
√ trust anchors are valid for at least 60 days
√ issuer cert is using supported crypto algorithm
√ issuer cert is within its validity period
√ issuer cert is valid for at least 60 days
√ issuer cert is issued by the trust anchor

linkerd-webhooks-and-apisvc-tls
-------------------------------
√ tap API server has valid cert
√ tap API server cert is valid for at least 60 days
√ proxy-injector webhook has valid cert
√ proxy-injector cert is valid for at least 60 days
√ sp-validator webhook has valid cert
√ sp-validator cert is valid for at least 60 days

linkerd-api
-----------
√ control plane pods are ready
√ control plane self-check
√ [kubernetes] control plane can talk to Kubernetes
√ [prometheus] control plane can talk to Prometheus
√ tap api service is running

linkerd-version
---------------
√ can determine the latest version
‼ cli is up-to-date
    is running version 2.9.1 but the latest stable version is 2.9.2
    see https://linkerd.io/checks/#l5d-version-cli for hints

control-plane-version
---------------------
‼ control plane is up-to-date
    is running version 2.9.1 but the latest stable version is 2.9.2
    see https://linkerd.io/checks/#l5d-version-control for hints
√ control plane and cli versions match

linkerd-prometheus
------------------
√ prometheus add-on service account exists
√ prometheus add-on config map exists
√ prometheus pod is running

linkerd-grafana
---------------
√ grafana add-on service account exists
√ grafana add-on config map exists
√ grafana pod is running

Status check results are √
```

### Environment

- Kubernetes Version: `Server Version: version.Info{Major:"1", Minor:"16+", GitVersion:"v1.16.15-gke.6000", GitCommit:"b02f5ea6726390a4b19d06fa9022981750af2bbc", GitTreeState:"clean", BuildDate:"2020-11-18T09:16:22Z", GoVersion:"go1.13.15b4", Compiler:"gc", Platform:"linux/amd64"}`
- Cluster Environment: GKE
- Host OS: Google Container-Optimized OS with Docker (cos)
- Linkerd version: stable-2.9.1 and edge-21.1.1

### Possible solution

Adding port 5432 to linkerd's skip ports is a workaround for now.

### Additional context

Slack discussion here: https://linkerd.slack.com/archives/C89RTCWJF/p1610480917116800

One more anecdote: I'm able to dump very small databases (and I suspect this is why normal queries in my app are working). Anything more than a couple of megabytes seems to trigger this hang.

**Comment by sjahl** [comment]
As requested, here's some TRACE level logs from a hanging pg_dump: https://gist.github.com/sjahl/f35175a9631de3f1a13f5d9692b0e6f1

I've removed some lines from the log (and noted where I did that). There was thousands of lines of logs relating to reading and writing data which I felt were less interesting, and may have caused some perf issues with github's gist. I left some of the original messages so that you can extrapolate what i've removed.

If the full log is desired, let me know, and I can share that in another venue.

**Comment by adleong** [comment]
Thanks for the detailed information @sjahl!  We will definitely look into this as we prepare for the stable-2.10 release.  Thanks for your patience!

**Comment by cpretzer** [comment]
@sjahl I have postgres deployed with a psql client and server in separate pods with the data that you've shared and I'm seeing the same behavior. I think the relevant lines in the output are:

`2021-01-27T20:29:40.872084407Z [   312.461937s] TRACE ThreadId(01) outbound: linkerd2_buffer::dispatch: Idling out inner service error=Service idled out after 5s
2021-01-27T20:29:40.872159125Z [   312.462185s] TRACE ThreadId(01) outbound: linkerd2_service_profiles::client: profile observation dropped

Next step is to look at the code to understand why this is deemed as an idle connection.

**Comment by cpretzer** [comment]
I've been able to reproduce this with stable-2.9.1 and edge-21.1.3. The interesting thing is that I see the `Service idled out after 5s` message in stable-2.9.1, but not in edge-21.1.3. [Here](https://gist.github.com/cpretzer/f804a5c8761bec577a0b5cbeddd525fd) is a full trace log from version edge-21.1.3 of the proxy

In addition, I found that the behavior occurs when loading the data using `psql` from Chapter 11 of the repository that @sjahl  mentioned above. I've also enabled the [Linkerd debug sidecar](https://linkerd.io/2/tasks/using-the-debug-container/) to capture packets.

I've created a [gist](https://gist.github.com/cpretzer/793b2b7da9975d06e04c6ed748c88571) with files and steps to repro. Note that this was run on GKE, so the PersistentVolume resources may be unpredictable in other environments.

1. Download the [gist files](https://gist.github.com/cpretzer/793b2b7da9975d06e04c6ed748c88571/archive/1a616c71166fcc2e74675ec7f83afed2981a3551.zip)
2. `unzip 1a616c71166fcc2e74675ec7f83afed2981a3551.zip`
3. Run `./setup.sh`
4. Run `./load.sh`

There are instructions printed from `load.sh` to reproduce the behavior. If at any time the containers restart, the `load.sh` script will do all the heavy lifting to get the env back into a reproducible state.

My next step is to test with the edge-21.1.4 to see if I can get more info.

**Comment by olix0r** [comment]
We should retest this on the most recent edge release. We've relaxed protocol detection to prevent failing connections, and I expect that this will resolve this issue.

**Comment by sjahl** [comment]
I'll see if I can find some time today to try the repro scripts on edge-21.2.2!

**Comment by sjahl** [comment]
OK, I set up a testing cluster and used the repro scripts from the gist, and it looks like I'm still seeing hangs with edge-21.2.2 :/

Behavior I observed:
  - with: `config.linkerd.io/opaque-ports: "5432"`: hangs
  - with no special port skips/opaque ports: hangs
  - with `config.linkerd.io/skip-inbound-ports: "5432"` and `config.linkerd.io/skip-outbound-ports: "5432"`: pg_dump completes successfully

For these tests, I was running pg_dump to stdout, but running the `psql -U linkerd -h postgres < /tmp/data.sql` as described results in the same hangs.

Linkerd debug logs for each case at: https://gist.github.com/sjahl/1f664f2c8566b9485ebe2119eecd8479

```
% linkerd check
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ controller pod is running
√ can initialize the client
√ can query the control plane API

linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
√ control plane CustomResourceDefinitions exist
√ control plane MutatingWebhookConfigurations exist
√ control plane ValidatingWebhookConfigurations exist
√ control plane PodSecurityPolicies exist

linkerd-identity
----------------
√ certificate config is valid
√ trust anchors are using supported crypto algorithm
√ trust anchors are within their validity period
√ trust anchors are valid for at least 60 days
√ issuer cert is using supported crypto algorithm
√ issuer cert is within its validity period
√ issuer cert is valid for at least 60 days
√ issuer cert is issued by the trust anchor

linkerd-webhooks-and-apisvc-tls
-------------------------------
√ proxy-injector webhook has valid cert
√ proxy-injector cert is valid for at least 60 days
√ sp-validator webhook has valid cert
√ sp-validator cert is valid for at least 60 days

linkerd-api
-----------
√ control plane pods are ready

linkerd-version
---------------
√ can determine the latest version
√ cli is up-to-date

control-plane-version
---------------------
√ control plane is up-to-date
√ control plane and cli versions match

Status check results are √

 % linkerd version
Client version: edge-21.2.2
Server version: edge-21.2.2
```

**Comment by olix0r** [comment]
@cpretzer the repro zip you shared didn't quite work for me out-of-the-box (namespace issues, etc); but I've modified it to work for me with locally [k3d](http://k3d.io):

1. `curl -sLO https://github.com/linkerd/linkerd2/files/5998724/gh-5539.tar.gz`
2. `tar -zxvf gh-5539.tar.gz && cd gh-5539`
3. `./create.sh` -- create a k3d cluster in docker with linkerd installed (expects `k3d` and `linkerd` to be in your `PATH`)
4. `./setup.sh`-- setup postgres and psql pods with volumes, etc
5. `./load.sh` -- copy data and run `psql -U linkerd -h postgres < /tmp/data.sql` 
6. `k3d cluster delete gh-5539` -- to cleanup when you're done

As far as I can tell, this all Just Works without issue for me (using edge-21.2.2). So, this points to some complexity introduced by the differing Kubernetes environment -- I'm not sure what, though.  It might be worth validating that my repro didn't remove anything important and that it works for you.... I'll try to figure out what we can do to narrow down this issue further.

**Comment by olix0r** [comment]
Is it possible that the Postgres instance is trying to initiate connections before the proxy has become available? This might explain why we don't see the problem in my single-node repro but do see it in other environments where the proxy may need to be pulled.

**Comment by sjahl** [comment]
The two data points that I can add here, which may help answer this:

1. I initially caught this problem when a nightly backup job for my database got stuck -- at this point, the pods that were involved in the transaction had been up and meshed for ~1/2 a day, so I don't think we were in a race with the proxy starting up.
2. In the logs for these failure cases, I can see `linkerd2_proxy_transport::connect` messages that indicate to me that it's successfully connecting through the proxy (could be reading this totally wrong, too 😅 )

I see that there was another edge release on the 26th. I'll give the repro case another shot with that -- will make sure that the proxies are pulled/up when I do that.

**Comment by sjahl** [comment]
Just tried a test on GKE with edge-21.2.4, and still hanging unfortunately. Trace logs (trimmed from the start of the postgres connection, up to around 10 seconds after the hang occurs), and an updated create.sh script which spins up a GKE cluster instead of k3d are here: https://gist.github.com/sjahl/6345063849a8cf69b5a2048d9d47dfad

Otherwise, all the other setup, scripts, and k8s manifests were exactly the same as what @olix0r posted in https://github.com/linkerd/linkerd2/files/5998724/gh-5539.tar.gz

**Comment by sjahl** [comment]
And, for completeness -- things also work flawlessly for me in a local k3d env.

**Comment by olix0r** [comment]
@sjahl Thanks, this is all super helpful context.

Let me share some context on what I see in the logs you shared:

First we see a connection received on 4143 -- meaning the peer has a proxy and discovered, via the control plane, that the communication is "opaque".
```
[   245.760815s] TRACE ThreadId(01) inbound: linkerd_proxy_transport::listen: Accepted server.addr=10.8.1.7:4143 client.addr=10.8.1.8:34942 orig.addr=Some(OrigDstAddr(10.8.1.7:4143))
[   245.760867s] DEBUG ThreadId(01) inbound: linkerd_app_inbound::prevent_loop: addr=10.8.1.7:4143 self.port=4143
[   245.761083s] DEBUG ThreadId(01) inbound:accept{client.addr=10.8.1.8:34942 target.addr=10.8.1.7:4143}: linkerd_tls::server: Peeked bytes from TCP stream sz=304
[   245.761110s] TRACE ThreadId(01) inbound:accept{client.addr=10.8.1.8:34942 target.addr=10.8.1.7:4143}: linkerd_tls::server::client_hello: parse_sni: parsed correctly up to SNI sni=Some(ServerId("default.gh-5539.serviceaccount.identity.linkerd.cluster.local"))
[   245.761121s] TRACE ThreadId(01) inbound:accept{client.addr=10.8.1.8:34942 target.addr=10.8.1.7:4143}: linkerd_tls::server: Identified matching SNI via peek sni=default.gh-5539.serviceaccount.identity.linkerd.cluster.local
[   245.764263s] DEBUG ThreadId(01) inbound:accept{client.addr=10.8.1.8:34942 target.addr=10.8.1.7:4143}: linkerd_tls::server: Accepted TLS connection client.id=Some(ClientId("default.gh-5539.serviceaccount.identity.linkerd.cluster.local")) alpn=Some("transport.l5d.io/v1")
```

We've used TLS ALPN to negotiate that both the client and server support the Linkerd "transport header", which indicates that the actual destination port is 5432. This is all necessary so that we can transport the stream over mTLS:
```
[   245.765293s] TRACE ThreadId(01) inbound:accept{client.addr=10.8.1.8:34942 target.addr=10.8.1.7:4143}:direct: linkerd_transport_header::server: Reading transport header
[   245.765480s] DEBUG ThreadId(01) inbound:accept{client.addr=10.8.1.8:34942 target.addr=10.8.1.7:4143}:direct: linkerd_transport_header::server: Read transport header header=TransportHeader { port: 5432, name: None, protocol: None }
```

We see a connection to 127.0.0.1:5432 as we'd expect:
```
[   245.765802s] DEBUG ThreadId(01) inbound:accept{client.addr=10.8.1.8:34942 target.addr=10.8.1.7:4143}:direct:opaque:tcp: linkerd_proxy_transport::connect: Connecting server.addr=127.0.0.1:5432
[   245.765969s] DEBUG ThreadId(01) inbound:accept{client.addr=10.8.1.8:34942 target.addr=10.8.1.7:4143}:direct:opaque:tcp: linkerd_proxy_transport::connect: Connected local.addr=127.0.0.1:35650 keepalive=None
```

And then we're off to the races proxying the TCP streams:
```
[   245.766108s] TRACE ThreadId(01) inbound:accept{client.addr=10.8.1.8:34942 target.addr=10.8.1.7:4143}:direct:opaque:tcp: linkerd_duplex: reading
[   245.766116s] TRACE ThreadId(01) inbound:accept{client.addr=10.8.1.8:34942 target.addr=10.8.1.7:4143}:direct:opaque:tcp: linkerd_duplex: read 8B
[   245.766121s] TRACE ThreadId(01) inbound:accept{client.addr=10.8.1.8:34942 target.addr=10.8.1.7:4143}:direct:opaque:tcp: linkerd_duplex: writing 8B
[   245.766198s] TRACE ThreadId(01) inbound:accept{client.addr=10.8.1.8:34942 target.addr=10.8.1.7:4143}:direct:opaque:tcp: linkerd_duplex: wrote 8B
[   245.766210s] TRACE ThreadId(01) inbound:accept{client.addr=10.8.1.8:34942 target.addr=10.8.1.7:4143}:direct:opaque:tcp: linkerd_duplex: reading
[   245.766218s] TRACE ThreadId(01) inbound:accept{client.addr=10.8.1.8:34942 target.addr=10.8.1.7:4143}:direct:opaque:tcp: linkerd_duplex: reading
[   245.768422s] TRACE ThreadId(01) inbound:accept{client.addr=10.8.1.8:34942 target.addr=10.8.1.7:4143}:direct:opaque:tcp: linkerd_duplex: poll
[   245.768481s] TRACE ThreadId(01) inbound:accept{client.addr=10.8.1.8:34942 target.addr=10.8.1.7:4143}:direct:opaque:tcp: linkerd_duplex: reading
[   245.768512s] TRACE ThreadId(01) inbound:accept{client.addr=10.8.1.8:34942 target.addr=10.8.1.7:4143}:direct:opaque:tcp: linkerd_duplex: reading
[   245.768560s] TRACE ThreadId(01) inbound:accept{client.addr=10.8.1.8:34942 target.addr=10.8.1.7:4143}:direct:opaque:tcp: linkerd_duplex: read 1B
...
```

And then the last message we see on the stream indicates that we're waiting for data in both directions:
```
[   245.817821s] TRACE ThreadId(01) inbound:accept{client.addr=10.8.1.8:34942 target.addr=10.8.1.7:4143}:direct:opaque:tcp: linkerd_duplex: reading
[   245.818565s] TRACE ThreadId(01) inbound:accept{client.addr=10.8.1.8:34942 target.addr=10.8.1.7:4143}:direct:opaque:tcp: linkerd_duplex: poll
[   245.818577s] TRACE ThreadId(01) inbound:accept{client.addr=10.8.1.8:34942 target.addr=10.8.1.7:4143}:direct:opaque:tcp: linkerd_duplex: reading
[   245.818584s] TRACE ThreadId(01) inbound:accept{client.addr=10.8.1.8:34942 target.addr=10.8.1.7:4143}:direct:opaque:tcp: linkerd_duplex: reading
```

So, from this proxy's point of view, at least, everything is healthy and normal and we're just waiting for more data... We may want to focus more on the client-side proxy. It may also be helpful to look at `ksniff` PCAPs, though at this point I'm not convinced they'll shine much light on the situation.

Thanks for all the help, @sjahl! I don't think we'll be able to nail this down before stable-2.10.0, but we should definitely continue to dig into this to figure out if we can reproduce this outside of gke and how these cases differ from the single-node k3d case.

**Comment by sjahl** [comment]
👍 Thanks for looking into this, I really appreciate it. No worries on the 2.10 release, I think keeping port 5432 skipped is reasonable for our project in the short/medium term.

I happened to have some Azure credits, so I ran this test there too, using their default 'Kubenet' network plugin, and getting some notably different behavior there: The initial load.sh succeeds! But, if I run it multiple times, or try other larger queries after the data is loaded, it stalls in a similar fashion to the clusters on GKE (for example, after a successful load.sh, running a pg_dump or `select * from nyc_yellow_taxi_trips_2016_06_01 limit 5000;` both seem to get hung up for me).

Logs from the successful load and a subsequent stalled `select * from nyc_yellow_taxi_trips_2016_06_01 limit 5000;` query are at: https://gist.github.com/sjahl/a26f07202efb8e026b62a15bee6e78ed

Happy hunting! :)

**Comment by olix0r** [comment]
@sjahl We're pretty confident that linkerd/linkerd2-proxy#944 fixes this -- it turns out that we were not calling `flush` on the socket, which was basically fine when dealing with raw TCP sockets, but because rustls (our backing TLS implementation) does its own buffering, these calls were necessary to prevent data from getting stuck in the buffer. With this proxy change, we now ensure that flush is called as needed and, in our repro environment, we no longer see traffic get stuck.

We're extraordinarily grateful for the detailed bug report. We plan to release this fix in an edge release today and in a stable release in the next few days (barring any surprises). Thanks!

**Comment by olix0r** [comment]
This is fixed in edge-21.3.2. Closing for now, but please let us know if you see anything unexpected. Again, thanks for the help tracking this down.

**Comment by sjahl** [comment]
Good work, all! Just to close the loop, the repro cases no longer fail for me with edge-21.3.2, so this indeed looks fixed. I haven't deployed the new release to any real environments yet, but I'll update here if I see any problems.
